docs: remove recommendation for prometheus metric_version=2

reimda opened the pull request:
Remove recommendation for metric_version=2 and add a description of both versions so users can choose which is better for their use case.

powersj commented on plugins/inputs/prometheus/README.md:
```diff
@@ -104,6 +104,35 @@ in Prometheus format.
+The Example Outputs sections shows examples for both options.
+
+When using this plugin along with the prometheus_client output, use the same
+option in both to ensure metrics are round-tripped without modification.
```
Thanks for this. Content looks good - should this be at the top, not buried in here? Or a line in the config to direct users here to learn about the differences?

reimda replied:
It can't be closer to the top unless it's before the config section. I think the config section belongs higher up, but a pointer in the config is good idea

powersj commented on plugins/outputs/prometheus_client/README.md:
```diff
@@ -16,7 +16,7 @@ metrics on `/metrics` (default) to be polled by a Prometheus server.
   ##
   ##   example: metric_version = 1;
-  ##            metric_version = 2; recommended version
+  ##            metric_version = 2;
```
do we need a similar section about the differences?

reimda replied:
I added a pointer here too

srebhan commented on plugins/inputs/prometheus/README.md:
```diff
@@ -104,6 +100,35 @@ in Prometheus format.
+sparse). Denseness is a useful property for some outputs, including those that
+are more efficient with row-oriented data.
+
+`metric_version = 2` differs in a few ways. The prometheus metric name becomes a
+telegraf field key. Metrics hold more than one value and the field keys aren't
+generic. The resulting metrics are sparse, but for some outputs they may be
+easier to process or query, including those that are more efficient with
+column-oriented data. The telegraf metric name is the same for all metrics in
+the input instance. It can be set with the `name_override` setting and defaults
+to "prometheus". To have multiple metric names, you can use multiple instances
+of the plugin, each with its own `name_override`.
```
Should we add a comment about round-trip-ability of this metric version versus version 1?

reimda replied:
I didn't write much on round tripping because I didn't have exact details on what works and doesn't work for both versions.  The serializer docs (https://github.com/influxdata/telegraf/blob/master/plugins/serializers/prometheus/README.md) imply that only 2 can round trip, but the input and output only say to make sure you use the same option in input and output if you want to round trip.

I was thinking it would be nice to add unit tests for round tripping if they don't already exist. That way it would be clear exactly what the shortcomings are and we could add clearer docs. I didn't want to hold up this PR on adding comprehensive unit tests though.

srebhan replied:
Fair enough.

srebhan reviewed (approved):
Looks good to me.

popey commented:
Looks good to me! Thanks @reimda

wangyuan0916 commented:
I'm confused... Why it is removed from docs? Isn't metric_version = 1 deprecated ? I saw these sentences from official release document.  https://docs.influxdata.com/telegraf/v1.23/release-notes-changelog/#v113-2019-12-12

>  The Prometheus Format (prometheus) input plugin and Prometheus Client (prometheus_client) output have a new mapping to and from Telegraf metrics, which can be enabled by setting metric_version = 2. **The original mapping is deprecated.**

euclidsun commented:
And the recommendation for v2 has been removed as well, can anyone share more context why is that?
What is the plan for v1, is it going to be deprecated in next few releases or will it still be around for a long time?

- the v1 deprecation were there for 1.13 ~ 1.17, and removed on 1.18
- the v2 as recommendation were there for 1.13 ~ 1.22
- now it's going to remove the v2 recommendation on 1.23 as well per the above

reimda commented:
@euclidsun @wangyuan0916 We removed the recommendation for v2 because it produces sparse metrics that don't perform well for some use cases and some storage backends. Since v2 isn't superior in all cases for all users, we didn't want to recommend it in all cases.

We removed the deprecation for v1 because it doesn't produce sparse metrics so it can perform better in some cases. We're not planning to remove it at this point. 

We leave it up to users to choose what version is best for them. Both mappings are in use and at this point we expect to keep both.

euclidsun commented:
thanks @reimda for the context and background, it helps!
